core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -45,15 +47,22 @@
 
     public static final String NAME = "same_shard";
 
-    public static final String SAME_HOST_SETTING = "cluster.routing.allocation.same_shard.host";
+    public static final Setting<Boolean> CLUSTER_ROUTING_ALLOCATION_SAME_HOST_SETTING =
+        Setting.boolSetting("cluster.routing.allocation.same_shard.host", false,
+            Setting.Property.Dynamic, Setting.Property.NodeScope);
```

Comment:
As this setting should usually be only set once, it is probably simpler to leave it non-dynamic (as @jasontedor suggested and as it was before this PR). In case where this must absolutely be updated on a production cluster, rolling restart (of master nodes) with config update is always possible.